src/vm/bifs.rs

```diff
@@ -4,48 +4,115 @@
 // http://www.apache.org/licenses/LICENSE-2.0 or the MIT license, see
 // LICENSE-MIT or http://opensource.org/licenses/MIT
 
+use crate::primitives::{Hash256, Hash512};
 use crate::vm::internal::VmTerm as Term;
 
+use blake2::digest::{Update, VariableOutput};
+use blake2::{Blake2bVar, Blake2s, Blake2sVar, Digest as Blake2Digest};
+use ripemd::{Digest as RipemdDigest, Ripemd160};
+use sha2::{Digest as ShaDigest, Sha256, Sha512};
+use sha3::{Digest as KeccakDigest, Keccak256, Keccak512};
+
 pub fn ripemd160(term: &Term) -> Term {
-    unimplemented!()
+    let mut hasher = Ripemd160::new();
+    RipemdDigest::update(&mut hasher, term.to_bytes_raw());
+    let hashed_term = hasher.finalize();
+
+    Term::Hash160(hashed_term.into())
 }
 
 pub fn sha256(term: &Term) -> Term {
-    unimplemented!()
+    let mut hasher = Sha256::new();
+    ShaDigest::update(&mut hasher, term.to_bytes_raw());
+    let hashed_term = hasher.finalize();
+
+    Term::Hash256(hashed_term.into())
 }
 
 pub fn sha512(term: &Term) -> Term {
-    unimplemented!()
+    let mut hasher = Sha512::new();
+    ShaDigest::update(&mut hasher, term.to_bytes_raw());
+    let hashed_term = hasher.finalize();
+
+    Term::Hash512(hashed_term.into())
 }
 
 pub fn keccak256(term: &Term) -> Term {
-    unimplemented!()
+    let mut hasher = Keccak256::new();
+    KeccakDigest::update(&mut hasher, term.to_bytes_raw());
+    let hashed_term = hasher.finalize();
+
+    Term::Hash256(hashed_term.into())
 }
 
 pub fn keccak512(term: &Term) -> Term {
-    unimplemented!()
+    let mut hasher = Keccak512::new();
+    KeccakDigest::update(&mut hasher, term.to_bytes_raw());
+    let hashed_term = hasher.finalize();
+
+    Term::Hash512(hashed_term.into())
 }
 
 pub fn blake2b_256(term: &Term) -> Term {
-    unimplemented!()
+    let mut hasher = Blake2bVar::new(32).unwrap();
+    hasher.update(&term.to_bytes_raw());
+    let hashed_term = hasher.finalize_boxed();
+
+    Term::Hash256(hashed_term.to_vec().try_into().unwrap())
 }
 
 pub fn blake2b_512(term: &Term) -> Term {
-    unimplemented!()
+    let mut hasher = Blake2bVar::new(64).unwrap();
+    hasher.update(&term.to_bytes_raw());
+    let hashed_term = hasher.finalize_boxed();
+
+    Term::Hash512(hashed_term.to_vec().try_into().unwrap())
 }
 
 pub fn blake2s_256(term: &Term) -> Term {
-    unimplemented!()
+    let mut hasher = Blake2sVar::new(32).unwrap();
+    hasher.update(&term.to_bytes_raw());
+    let hashed_term = hasher.finalize_boxed();
+
+    Term::Hash256(hashed_term.to_vec().try_into().unwrap())
 }
 
 pub fn blake2s_512(term: &Term) -> Term {
```

Comment:
This isn't needed and there's no opcode for it. Blake2s should always hash to 256 bits.

Reply:
modified in [fb60434](https://github.com/purpleprotocol/purplecoin/pull/30/commits/fb6043496abbc54334458fc15f2e56dcc8c8a890)